src/sai/i2s.rs

```diff
@@ -0,0 +1,748 @@
+    mode: I2SMode,
+    config: &I2SChanConfig,
+    mclk_div: u8,
+    bit_rate: I2SBitRate,
+) {
+    let clock_strobe = match config.clock_strobe {
+        I2SClockStrobe::Rising => false,
+        I2SClockStrobe::Falling => true,
+    };
+
+    // 16 bits in register correspond to 1 slot each max 16 slots
+    let slot_en_bits: u16 = (2_u32.pow(config.slots.into()) - 1) as u16;
+
+    let (frame_length, slot_size) = match bit_rate {
+        I2SBitRate::BITS_8 => (16 * (config.slots / 2), I2SSlotSize::BITS_16),
+        I2SBitRate::BITS_10 => (32 * (config.slots / 2), I2SSlotSize::BITS_16),
+        I2SBitRate::BITS_16 => (32 * (config.slots / 2), I2SSlotSize::BITS_16),
+        I2SBitRate::BITS_20 => (64 * (config.slots / 2), I2SSlotSize::BITS_32),
+        I2SBitRate::BITS_24 => (64 * (config.slots / 2), I2SSlotSize::BITS_32),
+        I2SBitRate::BITS_32 => (64 * (config.slots / 2), I2SSlotSize::BITS_32),
+    };
+
+    let mode_bits = (mode as u8) | (config.dir as u8);
+    unsafe {
+        audio_ch.cr1.modify(|_, w| {
+            w.mode()
+                .bits(mode_bits as u8)
+                .prtcfg()
+                .free()
+                .ds()
+                .bits(bit_rate as u8)
+                .lsbfirst()
+                .msb_first()
+                .ckstr()
+                .bit(clock_strobe)
+                .syncen()
+                .bits(config.sync_type as u8)
+                .mono()
+                .stereo()
+                .nodiv()
+                .master_clock()
+                .mckdiv()
+                .bits(mclk_div)
+        });
+        audio_ch.cr2.modify(|_, w| {
+            w.fth()
+                .quarter1()
+                .tris()
+                .clear_bit()
+                .mute()
+                .clear_bit()
+                .muteval()
+                .clear_bit()
+                .muteval()
+                .clear_bit()
+                .cpl()
+                .clear_bit()
+                // CPL, unused with companding off
+                .comp()
+                .no_companding()
+        });
+        audio_ch.frcr.modify(|_, w| {
+            w.frl()
+                .bits(frame_length - 1)
+                .fsall()
+                .bits((frame_length / 2) - 1)
+                .fsdef()
+                .set_bit() // left/right channels enabled
+                .fspol()
+                .set_bit() // FS active high
+                .fsoff()
+                .clear_bit()
+        });
+        audio_ch.slotr.modify(|_, w| {
+            w.fboff()
+                .bits(config.first_bit_offset)
+                .slotsz()
+                .bits(slot_size as u8)
+                .nbslot()
+                .bits(config.slots - 1)
+                .sloten()
+                .bits(slot_en_bits)
+        });
+        match mode {
+            I2SMode::Master => audio_ch
+                .im
+                .modify(|_, w| w.ovrudrie().enabled().wckcfgie().enabled()),
```

Comment:
For clarification it isn't needed in this passthrough example, because the `FREQIE` is set on the slave channel. However I think if you were using one (master) channel only, then it is needed if you want to feed/send data from an interrupt.

Reply:
All the interrupts should be accessible from the listen/unlisten methods now.